File: Difficulty: Medium/Count the number of possible triangles/count-the-number-of-possible-triangles.py
#User function Template for python3

class Solution:
    #Function to count the number of possible triangles.
    def countTriangles(self, arr):
        # code here
        arr.sort()
        n = len(arr)
        count = 0

        for i in range(n - 1, 1, -1):
            left = 0
            right = i - 1

            while left < right:
                if arr[left] + arr[right] > arr[i]:
                    count += right - left
                    right -= 1
                else:
                    left += 1

        return count


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        arr = list(map(int, input().strip().split()))
        ob = Solution()
        print(ob.countTriangles(arr))

        print("~")

# } Driver Code Ends

File: Difficulty: Medium/Count the number of possible triangles/count-the-number-of-possible-triangles.cpp
class Solution {
  public:
    int countTriangles(vector<int>& arr) {
    int n = arr.size();
    sort(arr.begin(), arr.end());
    int count = 0;

    for (int k = n - 1; k >= 2; k--) {
        int i = 0, j = k - 1;
        while (i < j) {
            if (arr[i] + arr[j] > arr[k]) {
                count += (j - i); // all elements between i and j are valid
                j--;
            } else {
                i++;
            }
        }
    }
    return count;
    }
};
